trainers/nodes/concept_extraction/sequence_to_sequence_ordered_concept_extraction/concept_extractor.py

```diff
@@ -0,0 +1,423 @@
+import dynet_config
+
+dynet_config.set(mem=1024)
+
+import dynet as dy
+
+from trainers.nodes.concept_extraction.sequence_to_sequence_ordered_concept_extraction.training_concepts_data_extractor \
+    import generate_concepts_training_data, ConceptsTrainingEntry
+from trainers.arcs.head_selection.head_selection_on_ordered_concepts.trainer_util import get_all_paths, \
+    plot_train_test_acc_loss
+from trainers.arcs.head_selection.head_selection_on_ordered_concepts.trainer import get_all_concepts
+
+from deep_dynet import support as ds
+from deep_dynet.support import Vocab
+
+from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
+
+bleu_smoothing = SmoothingFunction()
+
+# TODO?:
+# - Personalize preprocessing (consider the input that comes)
+# - Revise model
+# - Use tag info?
+
+EOS = "<EOS>"
+
+# SMALLER EMBEDDINGS AND STATE
+LSTM_NUM_OF_LAYERS = 2
+WORDS_EMBEDDING_SIZE = 50
+CONCEPTS_EMBEDDING_SIZE = 50
+STATE_SIZE = 50
+ATTENTION_SIZE = 50
+
+'''
+dyparams = dy.DynetParams()
+dyparams.set_mem(2048)
+dyparams.init()
+'''
+
+
+class ConceptsDynetGraph:
+    def __init__(self, words_vocab, concepts_vocab):
+        self.model = dy.Model()
+        self.words_vocab: Vocab = words_vocab
+        self.concepts_vocab: Vocab = concepts_vocab
+
+        self.input_embeddings = self.model.add_lookup_parameters((words_vocab.size(), WORDS_EMBEDDING_SIZE))
+        self.output_embeddings = self.model.add_lookup_parameters((concepts_vocab.size(), CONCEPTS_EMBEDDING_SIZE))
+
+        self.trainer = dy.SimpleSGDTrainer(self.model)
+
+        self.enc_fwd_lstm = dy.LSTMBuilder(LSTM_NUM_OF_LAYERS, WORDS_EMBEDDING_SIZE, STATE_SIZE, self.model)
+        self.enc_bwd_lstm = dy.LSTMBuilder(LSTM_NUM_OF_LAYERS, WORDS_EMBEDDING_SIZE, STATE_SIZE, self.model)
+
+        self.dec_lstm = dy.LSTMBuilder(LSTM_NUM_OF_LAYERS, STATE_SIZE * 2 + CONCEPTS_EMBEDDING_SIZE, STATE_SIZE,
+                                       self.model)
+
+        self.attention_w1 = self.model.add_parameters((ATTENTION_SIZE, STATE_SIZE * 2))
+        self.attention_w2 = self.model.add_parameters((ATTENTION_SIZE, STATE_SIZE * LSTM_NUM_OF_LAYERS * 2))
+        self.attention_v = self.model.add_parameters((1, ATTENTION_SIZE))
+
+        self.decoder_w = self.model.add_parameters((concepts_vocab.size(), STATE_SIZE))
+        self.decoder_b = self.model.add_parameters((concepts_vocab.size()))
+
+
+def embed_sequence(concepts_dynet_graph, sequence):
+    # Sentence already comes as list of words
+    sequence = list(sequence) + [EOS]
+
+    sequence = [concepts_dynet_graph.words_vocab.w2i[word] for word in sequence]
+    return [concepts_dynet_graph.input_embeddings[index] for index in sequence]
+
+
+def run_lstm(init_state, input_vecs):
+    s = init_state
+
+    out_vectors = []
+    for vector in input_vecs:
+        s = s.add_input(vector)
+        out_vector = s.output()
+        out_vectors.append(out_vector)
+    return out_vectors
```

Comment:
isn't this the same as calling transduce? wouldn't that be more efficient and you could get rid of this function?